### PR TITLE
Issue 2836 fixture collection bug

### PR DESCRIPTION
**Pull request by tom-dalton-fanduel**
Attempted fix for #2836.

The existing handling of 'nodeid' when determining what fixtures are visible is overly simplistic - it considers 'tests/food' to be a 'child' node of 'tests/foo', when in fact they are sibling nodes. This causes fixtures from 'tests/foo/conftest.py' to be used in tests like 'tests/food/test_food.py'.

This PR makes the hanlding more sophisticated - splitting the node id into components and then ensuring the base node id parts fully match the target node id's initial parts. This is still a little imperfect, in that would treat `foo/bar::Baz::()` as if it was the same as `foo/bar/Baz/()`. That seems a somewhat unlikely situation...

I'm not super keen on having `_splitnode` and `_ischildnode` as members of FixtureManager, but I don't really see any obvious other place for them and I was reluctant to create a while new module/class for these two 'nodeid' helper functions. I'm not 100% sure where the spec/definiton of a node id comes from, so maybe there is a better place, very open to suggestions on that. If someone with some better knowledge of what 'nodeid's look like can have a look over this it would be good - hopefulyl I haven't missed any edge cases but without having a good definition I am slightly worried about it.

The fixtures module didn't (seem to) have any existing unit tests, so I've added one for my new  `FixtureManager_ischildnode` method (mainly for my own sanity). I've also added a more 'integration'-y test to `test_collection.py` which will catch regression and (hopefully) follows the style of other tests there too. Hopefully `test_collection` is the right place for that test - certainly the fixture code I've modified seems to be run as part of the collection phase.

### CODE REVIEW

File: _pytest/fixtures.py [hunk 1]
```diff
@@ -1130,7 +1130,40 @@ def getfixturedefs(self, argname, nodeid):
         else:
             return tuple(self._matchfactories(fixturedefs, nodeid))
 
+    @classmethod
+    def _splitnode(cls, nodeid):
```
Comment (nimbol) [inline _pytest/fixtures.py]: It appears that this can be a `staticmethod` instead. The class isn't referenced anywhere in this method.

File: _pytest/fixtures.py [hunk 2]
```diff
@@ -1130,7 +1130,40 @@ def getfixturedefs(self, argname, nodeid):
...
+        """
+        if nodeid == '':
+            return []
+        sep = py.path.local.sep
+        parts = nodeid.split(sep)
+        last_part = parts[-1]
+        if '::' in last_part:
+            # Replace single last element 'test_foo.py::Bar::()' with multiple elements 'test_foo.py', 'Bar', '()'
+            parts[-1:] = last_part.split("::")
+        return parts
```
Comment (nimbol) [inline _pytest/fixtures.py]: Just a suggestion to simplify the code: remove the special case for the empty `nodeid`. The following would return functionally equivalent results (it would just return `['']` instead of `[]`) without special-casing anything:
```
sep = py.path.local.sep
parts = nodeid.split(sep)
parts[-1:] = parts[-1].split("::")
return parts
```

Comment (tom-dalton-fanduel) [reply _pytest/fixtures.py]: Thanks, I've shortened this down a bit (and made the docstring longer to compensate ;-))

File: _pytest/fixtures.py (hunk 1 above)
Comment (tom-dalton-fanduel) [reply _pytest/fixtures.py]: I'm not a fan of staticmethods, though mainly for inheritance related reasons and doubt anyone is subclassing fixtruemanager, let alone doing so and then wanting to change these. It feels like they should probably be plain functions in a 'node' module somewhere but hopefully someone more familiar with pytest can guide me on that... (see also my comment about `_splitnode` in the description.

File: changelog/2836.bug [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Attempted fix for https://github.com/pytest-dev/pytest/issues/2836
```
Comment (RonnyPfannschmidt) [inline changelog/2836.bug]: i propose `match fixture paths against actual path segments in order to avoid matching folders which share a prefix`

CC @nicoddemus

File: _pytest/fixtures.py (hunk 1 above)
Comment (RonnyPfannschmidt) [reply _pytest/fixtures.py]: since both helpers are completely unrelated to the class, how about just moving them out of the class instead

**Comment by tom-dalton-fanduel** [comment]
Thanks a lot for the feedback - I've pushed changes as suggested. Have put the node functions in a new module but let me know if you'd rather they remained in `fixtures.py`.

**Comment by RonnyPfannschmidt** [comment]
the linting issues are newly found issues we need to addto the ignore list
the xdist failure is a flaky test we are aware of so this should be mergable

CC @nicoddemus

**Comment by nicoddemus** [comment]
Thanks a lot @tom-dalton-fanduel for the PR, we appreciate it!

I have added the two flake8 new exceptions to the ignore list (E722: "do not use bare except" and E741: "ambiguous variable name 'l'") and fixed another one.

After CI passes I we can merge this. 👍

**Comment by nicoddemus** [comment]
Tests are failing on Windows, I suspect we also need to split by [`os.altsep`](https://docs.python.org/3/library/os.html#os.altsep) when not `None`.

File: _pytest/nodes.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    Return values are lists e.g.
+        []
+        ['testing', 'code']
+        ['testing', 'code', 'test_excinfo.py']
+        ['testing', 'code', 'test_excinfo.py', 'TestFormattedExcinfo', '()']
+    """
+    if nodeid == '':
+        # If there is no root node at all, return an empty list so the caller's logic can remain sane
+        return []
+    parts = nodeid.split(py.path.local.sep)
```
Comment (RonnyPfannschmidt) [inline _pytest/nodes.py]: https://github.com/pytest-dev/pytest/blob/master/_pytest/main.py#L519-L527 seems to generally replace the path separator with '/' for consistent node ids across operating systems, i think introducing a constant for that is necessary

Comment (tom-dalton-fanduel) [reply _pytest/nodes.py]: I've created a constant `SEP` in the nodes module, and used that in all the places I could see that used a literal `'/'` in the context of node ids. I subsequently noticed there is a `Node` class in main.py (I hadn't noticed that before now), but I can't see an (easy) way to move the new plain functions into that class, since they're used with strings rather than actual node objects.

**Comment by tom-dalton-fanduel** [comment]
@nicoddemus is the failing Windows build visible (to me) anywhere? I don't see anything obviously windows-y in Travis...

**Comment by RonnyPfannschmidt** [comment]
this now is a beautiful cleanup 👍  - the  ci issues are from a flake8 update and a bug that @nicoddemus  is currently working on

**Comment by RonnyPfannschmidt** [comment]
merging as the failures are unrelated, good work 👍

**Comment by tom-dalton-fanduel** [comment]
Party Parrot

**Comment by nicoddemus** [comment]
Hahaha that's an awesome gif! 😄

**Comment by tom-dalton-fanduel** [comment]
Yep - while searching for an appropriate "I am officially a contibutor to pytest, hooray" gif, I came across [Party Parrot as a Service](https://parrotify.github.io/) and couldn't resist!

**Comment by nicoddemus** [comment]
I can say with all certainty that this is one of the most useful services of all time. 😆
